python/frontend/cirrus/cirrus/lambdas/handler.py

```diff
@@ -0,0 +1,164 @@
+
+    return redis_client, node_manager
+
+
+def feature_hashing_handler(s3_client, data, labels, event):
+    """ Handle a call for feature hashing """
+    timer = Timer("CHUNK{0}".format(event["s3_key"])).set_step(
+        "Feature hashing")
+    printer = prefix_print("CHUNK{0}".format(event["s3_key"]))
+    printer("Hashing data")
+    hashed = feature_hashing_helper.hash_data(data,
+                                              event["columns"],
+                                              event["n_buckets"])
+    printer("Serializing data")
+    serialized = serialize_data(hashed, labels)
+    printer("Putting object in S3")
+    s3_client.put_object(
+        Bucket=event["s3_bucket_output"], Key=event["s3_key"], Body=serialized)
+    timer.timestamp()
+
+
+def min_max_handler(s3_client, redis_client, data, labels,
+                    node_manager, redis_flag, event):
+    """ Either calculates the local bounds, or scales data and puts
+    the new data in {src_object}_scaled. """
+    timer = Timer("CHUNK{0}".format(event["s3_key"]))
+    if event["action"] == "LOCAL_BOUNDS":
+        print("Getting local data bounds...")
+        timer.set_step("Calculating bounds")
+        bounds = min_max_helper.get_data_bounds(data)
+        timer.timestamp().set_step("Putting bounds in S3 / Redis")
+        print("Putting bounds in S3...")
+        min_max_helper.put_bounds_in_db(
+            s3_client, redis_client, bounds,
+            event["s3_bucket_input"], event["s3_key"] + "_bounds",
+            redis_flag, node_manager, event["s3_key"])
+        timer.timestamp()
+    elif event["action"] == "LOCAL_SCALE":
+        assert "s3_bucket_output" in event, "Must specify output bucket."
+        assert "min_v" in event, "Must specify min."
+        assert "max_v" in event, "Must specify max."
+        print("Getting global bounds...")
+        timer.set_step("Getting global bounds")
+        bounds = min_max_helper.get_global_bounds(
+            s3_client, redis_client, event["s3_bucket_input"], event["s3_key"],
+            redis_flag, event["s3_key"])
+        timer.timestamp().set_step("Scaling data")
+        print("Scaling data...")
+        scaled = min_max_helper.scale_data(data, bounds,
+                                           event["min_v"], event["max_v"])
+        timer.timestamp().set_step("Serializing")
+        print("Serializing...")
+        serialized = serialize_data(scaled, labels)
+        timer.timestamp().set_step("Putting in S3")
+        s3_client.put_object(
+            Bucket=event["s3_bucket_output"],
+            Key=event["s3_key"], Body=serialized)
+        timer.timestamp()
+
+
+def normal_scaling_handler(s3_client, data, labels, event):
+    """ Scale to a unit normal range """
+    if event["action"] == "LOCAL_RANGE":
+        print("Getting local data ranges...")
+        bounds = normal_helper.get_data_ranges(data)
+        print("Putting ranges in S3...")
+        min_max_helper.put_bounds_in_db(
```

Comment:
I may be wrong but I don't think this code will work. put_bounds_in_db calls get_keys_values which assumes the input data has a "max" key. normal_helper.get_data_ranges(data) doesn't seem to comply with this.